lib/odbc_adapter/database_statements.rb

```diff
@@ -8,19 +8,28 @@ module DatabaseStatements
     # Executes the SQL statement in the context of this connection.
     # Returns the number of rows affected.
     def execute(sql, name = nil, binds = [])
+      sql = transform_query(sql)
       log(sql, name) do
         sql = bind_params(binds, sql) if prepared_statements
-        @connection.do(sql)
+        @raw_connection.do(sql)
       end
     end
 
-    # Executes +sql+ statement in the context of this connection using
-    # +binds+ as the bind substitutes. +name+ is logged along with
-    # the executed +sql+ statement.
-    def exec_query(sql, name = 'SQL', binds = [], prepare: false) # rubocop:disable Lint/UnusedMethodArgument
+    # Executes an INSERT query and returns the new record's ID
+    def insert(arel, name = nil, pk = nil, id_value = nil, sequence_name = nil, binds = [], returning: nil)
+      sql, binds = to_sql_and_binds(arel, binds)
+      exec_insert(sql, name, binds, pk, sequence_name, returning: returning)
+
+      return [id_value] unless returning.nil?
+
+      id_value
+    end
+
+    def internal_exec_query(sql, name = 'SQL', binds = [], prepare: false) # rubocop:disable Lint/UnusedMethodArgument
+      sql = transform_query(sql)
       log(sql, name) do
         sql = bind_params(binds, sql) if prepared_statements
-        stmt =  @connection.run(sql)
+        stmt =  @raw_connection.run(sql)
```

Comment:
Some popular active record adapter implementations(e.g. [mysql2](https://github.com/rails/rails/blob/19eebf6d33dd15a0172e3ed2481bec57a89a2404/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb#L23)) have their own [`raw_execute` definition](https://github.com/rails/rails/blob/19eebf6d33dd15a0172e3ed2481bec57a89a2404/activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb#L96-L106) and uses the [abstract adapter method with_raw_connection](https://github.com/rails/rails/blob/19eebf6d33dd15a0172e3ed2481bec57a89a2404/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L999) to run  statements.

Moreover, using with_raw_connection instead of accessing @raw_connection directly has some advantages:

**Encapsulation**: [with_raw_connection](https://github.com/rails/rails/blob/null/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L999-L1061) provides a controlled way to access the raw connection, ensuring that any necessary setup or teardown logic is executed. This encapsulation helps maintain the integrity of the connection.

**Error Handling**: It can include error handling mechanisms to manage exceptions that might occur during the use of the raw connection, providing a more robust and resilient code.

**Resource Management**: It can ensure that resources are properly managed, such as closing the connection after use, which helps prevent resource leaks.

**Consistency**: Using a method like with_raw_connection ensures that all access to the raw connection goes through a single point, making it easier to maintain and modify the connection handling logic in one place.

**Abstraction**: It abstracts the details of how the raw connection is managed, allowing for changes in the underlying implementation without affecting the rest of the codebase.


@zokioki Do you think that we could benefit from following the same approach as other popular active record adapter implementations?

Reply:
That's a good point @phrsouza, I agree that following a similar approach could be beneficial here. It might be a good exercise to sit down and look at a refactoring pass on this when we have some time. Echoing @austenmadden's comment above, it might also be worth it to see if there's anything we can do to centralize some of the maintenance efforts for this gem that are spread across a few different forks (e.g. [springbuk](https://github.com/springbuk/odbc_adapter), [patterninc](https://github.com/patterninc/odbc_adapter), [instacart](https://github.com/instacart/odbc_adapter)).